README.md
# Customer Success Demo: Weather Widgets - A Stencil Based Web Component Library

If you are looking for a library that will provide components that you can use to build the next great weather application, you are in the wrong spot.

This project was put together in order to show a simple monorepo architecture and build system. This project uses [Stencil](https://stenciljs.com/) to build a library of web components and then uses the [Stencil Framework Wrappers](https://github.com/ionic-team/stencil-ds-output-targets) to package the components for use with each of the three major application frameworks (Angular, React, and Vue).

## Building

If you would like to build this for yourself, you can follow these steps.

- `git clone`
- `npm i`
- `npm run bootstrap`
- `npm run build`

## Publishing

These packages are specifically scoped to our `@ionic-enterprise` scope. As such, you need to be a member of the Ionic Enterprise org in order to publish them. If you would like to publish your own version for any reason, please fork this project and change the package names to be scoped to your own user or organization. Once you have done that, you should be able to run `npm run release` in order to publish.

The `npm run release` command will ask how the version should be bumped, run a linting, run the tests, and run a build. If that all passes, the packages will then be published to NPM.

One note on the publishing. By default, NPM publishes scoped packages as private. This can only be done if you are paying money to NPM. You _can_, however, publish public scoped packages for free. If you look in the `package.json` for each individual package you will see the following configuration specifying to do exactly that:

```JSON
  "publishConfig": {
    "access": "public"
  },
```

## Building Your Own Library

You may want to create your own web component library using Stencil. For this reason, I will document here how I went about creating this library. This `README` just covers the basic tooling of the overall project. To see how each of the individual `packages` was created, please see the `README` for each package individually.

### The Monorepo

When creating a Stencil based web component library that uses the Ionic Framework wrappers, it makes a lot of sense to organize the projects into a monorepo structure. It is best to use some sort of dedicated tooling to manage the monorepo. For this repo I chose to use [Lerna](https://github.com/lerna/lerna) to provide the tooling and structure.

To set up the monorepo:

```
mkdir cs-demo-weather-widgets
cd cs-demo-weather-widgets
npx lerna init
```

### The `package.json` File

At this point the basic infrastructure is in place. Edit the `package.json` file to contain the information that is required as well as the scripts to use while developing the library.

Update the basic information (`name`, `description`, `author`, etc). Those values do not really mean much at this level in the repo, but they provide information to future developers.

The most important part is the `scripts` section. Keep the scripts as simple as possible while also providing the functionality that I need. For example:

```json
  "scripts": {
    "bootstrap": "lerna bootstrap",
    "build": "lerna run --scope @ionic-enterprise/cs-demo-weather-widgets build && lerna run --parallel --ignore @ionic-enterprise/cs-demo-weather-widgets build",
    "lint": "lerna run lint --parallel",
    "preversion": "npm run lint && npm run test && npm run build",
    "release": "lerna publish",
    "test": "lerna run --scope @ionic-enterprise/cs-demo-weather-widgets test"
  },
```

- **bootstrap**: Bootstrap each of the packages. That is, install their dependencies and create cross links between internal dependencies.
- **build**: First build the web component library. This generates code in the other packages. Then build the other packages.
- **lint**: Run lint in any package that has a `lint` command defined.
- **preversion**: Before bumping the version, make sure the current code lints, the tests pass, and the build works.
- **release**: Bump the version, run the `preversion` script, and publish to NPM.
- **test**: The only package that is tested is the `@ionic-enterprise/cs-demo-weather-widgets` package. All other code is generated, and as such it would be difficult to create comprehensive tests for them.

### The Lerna Configuration

This project uses the default `lerna.json` configuration file with minor modifications to support conventional commits.

```json
{
  "packages": ["packages/*"],
  "version": "1.1.1",
  "command": {
    "create": {
      "homepage": "https://github.com/ionic-enterprise/cs-demo-weather-widgets",
      "license": "MIT"
    },
    "version": {
      "allowBranch": "main",
      "conventionalCommits": true,
      "message": "chore(release): %s"
    }
  }
}
```

### Git Tooling

Code formatting wars and people pushing code that does not lint or has failing tests can be serious problems. [Git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) provide a great mechanism to avoid these kinds of issues through the enforce of standards.

Here is an example:

```
npm install -D @ionic/prettier-config husky prettier pretty-quick
```

[Husky](https://typicode.github.io/husky/#/) is used to manage the git hooks. [Prettier](https://prettier.io/) and [pretty-quick](https://www.npmjs.com/package/pretty-quick) are used to facilitate the code formatting according to [our standards](https://www.npmjs.com/package/@ionic/prettier-config).

Make the following changes to the `package.json`:

```JSON
{
  ...
  "scripts": {
    ...
    "prepare": "husky install",
    ...
  },
  ...
  "prettier": "@ionic/prettier-config"
}
```

To initially set up the git hooks, you can run the `prepare` script you just added. For other developers who clone your project, this happens automatically when they install the dependencies.

```
npm run prepare
```

This creates a `.husky` directory for your hooks. I create two hooks: `pre-commit` and `pre-push`.

```bash
npx husky add .husky/pre-commit "npx pretty-quick --staged"
npx husky add .husky/pre-push "npm run lint"
```

Update the `pre-push` to run the tests on a push to `main`:

```bash
#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

npm run lint

branch=`git rev-parse --abbrev-ref HEAD`
if [ "main" = "$branch" ]; then
  npm run test
fi;
```

Here is the reasoning behind the git-hooks:

1. Improperly formatted code should _never_ be committed. The `pre-commit` hook will fix any bad formatting.
1. Your code should always lint. Period. If it doesn't then you can't push it, even to your own feature branch.
1. Your code may have some testing issues while in development, but those absolutely need to be fixed before that code makes it into the main branch.

There is _one_ problem, though. This repo contains a lot of generated code, and we cannot be responsible for its formatting. Create a `.prettierignore` file to ignore the generated code.

```
packages/core/src/components.d.ts
packages/core/src/components/**/readme.md
packages/angular
packages/react
packages/vue
```

### The `.gitignore` File

Lerna does not include a `.gitignore` file, so we need to create our own. I use the following basic `.gitignore` in the root of the monorepo:

```
*~
*.sw[mnpcod]
*.log
*.lock
*.tmp
*.tmp.*
log.txt
*.sublime-project
*.sublime-workspace
.idea/
.vscode/
.sass-cache/
.versions/
node_modules/
$RECYCLE.BIN/

.DS_Store
Thumbs.db
UserInterfaceState.xcuserstate
.env
```

Other packages within the monorepo will have their own `.gitignore` files with other items that are specific to them.

### Packages

The following packages have been created. See the `NOTES.md` files for details on how they were created:

- **[core](packages/core/NOTES.md)**: the Stencil based web components, distributed on via NPM as `@ionic-enterprise/cs-demo-weather-widgets`
- **[angular](packages/angular/NOTES.md)**: Angular proxies, distributed via NPM as `@ionic-enterprise/cs-demo-weather-widgets angular`
- **[react](packages/react/NOTES.md)**: React proxies, distributed via NPM as `@ionic-enterprise/cs-demo-weather-widgets-react`
- **[vue](packages/vue/NOTES.md)**: Vue proxies, distributed via NPM as `@ionic-enterprise/cs-demo-weather-widgets-vue`

When creating your own library, start with the Stencil library (`core` in this case). With a simple library in place, create the projects for each framework wrapper. At that point, circle back to the Stencil library and continue to develop the web components.

Happy Coding!! 🤓


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.2.1](https://github.com/ionic-enterprise/cs-demo-weather-widgets/compare/v1.2.0...v1.2.1) (2022-04-22)

### Bug Fixes

- **vue:** make publish public ([49716e8](https://github.com/ionic-enterprise/cs-demo-weather-widgets/commit/49716e859ecfba8bba463184a75f45720ba34e2d))

# [1.2.0](https://github.com/ionic-enterprise/cs-demo-weather-widgets/compare/v1.1.1...v1.2.0) (2022-04-22)

### Bug Fixes

- **base:** remove junk orig folder ([e8ff881](https://github.com/ionic-enterprise/cs-demo-weather-widgets/commit/e8ff88190506928abce23ac8dedd930c5ef52f19))

### Features

- **angular:** add the angular framework wrapper ([7a40fe7](https://github.com/ionic-enterprise/cs-demo-weather-widgets/commit/7a40fe73b84eab1549cef2071ecdc78609b9ded1))
- **vue:** add the vue framework wrapper ([d058525](https://github.com/ionic-enterprise/cs-demo-weather-widgets/commit/d05852507be466dfb4829a183742aecc56d54992))

## [1.1.1](https://github.com/ionic-enterprise/cs-demo-weather-widgets/compare/v1.1.0...v1.1.1) (2022-04-22)

### Bug Fixes

- **react:** update the types path ([133a41e](https://github.com/ionic-enterprise/cs-demo-weather-widgets/commit/133a41eb513a8548bae7d81bbe37574beb1138a5))

<a name="1.0.1"></a>

## 1.0.1 (2019-01-23)

<a name="1.0.0"></a>

# [1.0.0](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.10-2...v1.0.0) (2018-11-29)

### Bug Fixes

- **tests:** convert some unit tests to e2e tests ([d0e0f5a](https://github.com/kensodemann/csdemo-weather-widgets/commit/d0e0f5a))

<a name="0.0.10-2"></a>

## [0.0.10-2](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.10-1...v0.0.10-2) (2018-11-27)

<a name="0.0.10-1"></a>

## [0.0.10-1](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.10-0...v0.0.10-1) (2018-11-27)

<a name="0.0.10-0"></a>

## [0.0.10-0](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.9...v0.0.10-0) (2018-11-27)

### Features

- **condition:** allow setting of the background color ([42416a9](https://github.com/kensodemann/csdemo-weather-widgets/commit/42416a9))
- **upgrade:** initial work to upgrade to latest stencil ([ed30636](https://github.com/kensodemann/csdemo-weather-widgets/commit/ed30636))

<a name="0.0.9"></a>

## [0.0.9](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.8...v0.0.9) (2018-07-09)

### Features

- **styles:** use custom properties ([4fec407](https://github.com/kensodemann/csdemo-weather-widgets/commit/4fec407))

<a name="0.0.8"></a>

## 0.0.8 (2018-07-02)

### Bug Fixes

- **forecast:** handle component load ([ff175f2](https://github.com/kensodemann/csdemo-weather-widgets/commit/ff175f2))

### Features

- **condition:** add the weather condition component ([3736c8e](https://github.com/kensodemann/csdemo-weather-widgets/commit/3736c8e))
- **css:** remove the shadow DOM setting ([61d922e](https://github.com/kensodemann/csdemo-weather-widgets/commit/61d922e))
- **forecast:** add a daily forecast component ([e704b6a](https://github.com/kensodemann/csdemo-weather-widgets/commit/e704b6a))
- **forecast:** add the date to the forecast ([29d250b](https://github.com/kensodemann/csdemo-weather-widgets/commit/29d250b))
- **temperature:** create a temperature component ([1da7bc1](https://github.com/kensodemann/csdemo-weather-widgets/commit/1da7bc1))
- **uv-index:** create UV Index widget ([db68a86](https://github.com/kensodemann/csdemo-weather-widgets/commit/db68a86))

<a name="0.0.7"></a>

## 0.0.7 (2018-06-12)

### Bug Fixes

- **forecast:** handle component load ([ff175f2](https://github.com/kensodemann/csdemo-weather-widgets/commit/ff175f2))

### Features

- **condition:** add the weather condition component ([3736c8e](https://github.com/kensodemann/csdemo-weather-widgets/commit/3736c8e))
- **css:** remove the shadow DOM setting ([61d922e](https://github.com/kensodemann/csdemo-weather-widgets/commit/61d922e))
- **forecast:** add a daily forecast component ([e704b6a](https://github.com/kensodemann/csdemo-weather-widgets/commit/e704b6a))
- **forecast:** add the date to the forecast ([29d250b](https://github.com/kensodemann/csdemo-weather-widgets/commit/29d250b))
- **temperature:** create a temperature component ([1da7bc1](https://github.com/kensodemann/csdemo-weather-widgets/commit/1da7bc1))
- **uv-index:** create UV Index widget ([db68a86](https://github.com/kensodemann/csdemo-weather-widgets/commit/db68a86))

<a name="0.0.6"></a>

## [0.0.6](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.5...v0.0.6) (2018-05-21)

### Bug Fixes

- **forecast:** handle component load ([ff175f2](https://github.com/kensodemann/csdemo-weather-widgets/commit/ff175f2))

### Features

- **forecast:** add the date to the forecast ([29d250b](https://github.com/kensodemann/csdemo-weather-widgets/commit/29d250b))

<a name="0.0.5"></a>

## [0.0.5](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.4...v0.0.5) (2018-05-15)

<a name="0.0.4"></a>

## [0.0.4](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.3...v0.0.4) (2018-05-15)

<a name="0.0.3"></a>

## [0.0.3](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.3-3...v0.0.3) (2018-05-15)

### Features

- **css:** remove the shadow DOM setting ([61d922e](https://github.com/kensodemann/csdemo-weather-widgets/commit/61d922e))

<a name="0.0.2"></a>

## [0.0.2](https://github.com/kensodemann/csdemo-weather-widgets/compare/v0.0.1...v0.0.2) (2018-05-15)

<a name="0.0.1"></a>

## 0.0.1 (2018-05-13)

### Features

- **condition:** add the weather condition component ([3736c8e](https://github.com/kensodemann/csdemo-weather-widgets/commit/3736c8e))
- **forecast:** add a daily forecast component ([e704b6a](https://github.com/kensodemann/csdemo-weather-widgets/commit/e704b6a))

src/services/weather-condition/weather-condition.spec.ts
import { WeatherCondition } from './weather-condition';
import { ConditionIconPaths } from '../../models/condition-icon-paths';

describe('Weather Condition Service', () => {
  let weatherCondition: WeatherCondition;
  beforeEach(() => {
    weatherCondition = new WeatherCondition();
  });

  it('builds', () => {
    expect(weatherCondition).toBeTruthy();
  });

  describe('rank', () => {
    it('is 0 for condition 810', () => {
      expect(weatherCondition.rank(810)).toEqual(0);
    });

    it('is 1 for condition 800', () => {
      expect(weatherCondition.rank(800)).toEqual(1);
    });

    it('is 1 for condition 904', () => {
      expect(weatherCondition.rank(904)).toEqual(1);
    });

    it('is 10 for 801 through 809', () => {
      for (let x = 801; x < 810; x++) {
        expect(weatherCondition.rank(x)).toEqual(10);
      }
    });

    it('is 20 for 300 through 399', () => {
      for (let x = 300; x < 400; x++) {
        expect(weatherCondition.rank(x)).toEqual(20);
      }
    });

    it('is 30 for 500 through 599', () => {
      for (let x = 500; x < 600; x++) {
        expect(weatherCondition.rank(x)).toEqual(30);
      }
    });

    it('is 40 for 701 through 771', () => {
      for (let x = 701; x < 772; x++) {
        expect(weatherCondition.rank(x)).toEqual(40);
      }
    });

    it('is 50 for 230 through 299', () => {
      for (let x = 230; x < 300; x++) {
        expect(weatherCondition.rank(x)).toEqual(50);
      }
    });

    it('is 60 for 200 through 229', () => {
      for (let x = 200; x < 230; x++) {
        expect(weatherCondition.rank(x)).toEqual(60);
      }
    });

    it('is 70 for 600 through 699', () => {
      for (let x = 600; x < 700; x++) {
        expect(weatherCondition.rank(x)).toEqual(70);
      }
    });

    it('is 70 for 903', () => {
      expect(weatherCondition.rank(903)).toEqual(70);
    });
  });

  describe('description', () => {
    it('is Unknown for condition 810', () => {
      expect(weatherCondition.description(810)).toEqual('Unknown');
    });

    it('is Sunny for condition 800', () => {
      expect(weatherCondition.description(800)).toEqual('Sunny');
    });

    it('is Sunny for condition 904', () => {
      expect(weatherCondition.description(904)).toEqual('Sunny');
    });

    it('is Cloudy for 801 through 809', () => {
      for (let x = 801; x < 810; x++) {
        expect(weatherCondition.description(x)).toEqual('Cloudy');
      }
    });

    it('is Rain for 300 through 399', () => {
      for (let x = 300; x < 400; x++) {
        expect(weatherCondition.description(x)).toEqual('Rain');
      }
    });

    it('is Showers for 500 through 599', () => {
      for (let x = 500; x < 600; x++) {
        expect(weatherCondition.description(x)).toEqual('Showers');
      }
    });

    it('is Fog for 701 through 771', () => {
      for (let x = 701; x < 772; x++) {
        expect(weatherCondition.description(x)).toEqual('Fog');
      }
    });

    it('is Scattered Storms for 230 through 299', () => {
      for (let x = 230; x < 300; x++) {
        expect(weatherCondition.description(x)).toEqual('Scattered Storms');
      }
    });

    it('is Thunderstorms for 200 through 229', () => {
      for (let x = 200; x < 230; x++) {
        expect(weatherCondition.description(x)).toEqual('Thunderstorms');
      }
    });

    it('is Snow for 600 through 699', () => {
      for (let x = 600; x < 700; x++) {
        expect(weatherCondition.description(x)).toEqual('Snow');
      }
    });

    it('is Snow for 903', () => {
      expect(weatherCondition.description(903)).toEqual('Snow');
    });
  });

  describe('image url', () => {
    let paths: ConditionIconPaths;
    beforeEach(() => {
      paths = {
        sunny: './assets/images/sunny.png',
        cloudy: './assets/images/cloudy.png',
        lightRain: './assets/images/light-rain.png',
        shower: './assets/images/shower.png',
        sunnyThunderStorm: './assets/images/sunny-tstorm.png',
        thunderStorm: './assets/images/tstorm.png',
        fog: './assets/images/fog.png',
        snow: './assets/images/snow.png',
        unknown: './assets/images/unknown.png'
      };
    });

    it('is null when no image urls exist', () => {
      expect(weatherCondition.imageUrl(800, null)).toEqual(null);
      expect(weatherCondition.imageUrl(800, undefined)).toEqual(null);
    });

    it('is Unknown for condition 810', () => {
      expect(weatherCondition.imageUrl(810, paths)).toEqual(paths.unknown);
    });

    it('is Sunny for condition 800', () => {
      expect(weatherCondition.imageUrl(800, paths)).toEqual(paths.sunny);
    });

    it('is Sunny for condition 904', () => {
      expect(weatherCondition.imageUrl(904, paths)).toEqual(paths.sunny);
    });

    it('is Cloudy for 801 through 809', () => {
      for (let x = 801; x < 810; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(paths.cloudy);
      }
    });

    it('is Rain for 300 through 399', () => {
      for (let x = 300; x < 400; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(paths.lightRain);
      }
    });

    it('is Showers for 500 through 599', () => {
      for (let x = 500; x < 600; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(paths.shower);
      }
    });

    it('is Fog for 701 through 771', () => {
      for (let x = 701; x < 772; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(paths.fog);
      }
    });

    it('is Scattered Storms for 230 through 299', () => {
      for (let x = 230; x < 300; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(
          paths.sunnyThunderStorm
        );
      }
    });

    it('is Thunderstorms for 200 through 229', () => {
      for (let x = 200; x < 230; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(paths.thunderStorm);
      }
    });

    it('is Snow for 600 through 699', () => {
      for (let x = 600; x < 700; x++) {
        expect(weatherCondition.imageUrl(x, paths)).toEqual(paths.snow);
      }
    });

    it('is Snow for 903', () => {
      expect(weatherCondition.imageUrl(903, paths)).toEqual(paths.snow);
    });
  });

  describe('high', () => {
    it('is undefined if there are no forecasts', () => {
      expect(weatherCondition.high(null)).toBeUndefined();
      expect(weatherCondition.high(undefined)).toBeUndefined();
      expect(weatherCondition.high([])).toBeUndefined();
    });

    it('returns the highest temperature', () => {
      expect(
        weatherCondition.high([
          {
            date: new Date(),
            condition: 800,
            temperature: 287.15
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 286.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 289.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 290.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 295.15
          },
          {
            date: new Date(),
            condition: 305,
            temperature: 297.15
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 295.65
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 293.65
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 288.65
          }
        ])
      ).toEqual(297.15);
    });
  });

  describe('low', () => {
    it('is undefined if there are no forecasts', () => {
      expect(weatherCondition.low(null)).toBeUndefined();
      expect(weatherCondition.low(undefined)).toBeUndefined();
      expect(weatherCondition.low([])).toBeUndefined();
    });

    it('returns the lowest temperature', () => {
      expect(
        weatherCondition.low([
          {
            date: new Date(),
            condition: 800,
            temperature: 287.15
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 286.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 289.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 290.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 295.15
          },
          {
            date: new Date(),
            condition: 305,
            temperature: 297.15
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 295.65
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 293.65
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 288.65
          }
        ])
      ).toEqual(286.65);
    });
  });

  describe('most serious condition', () => {
    it('is undefined if there are no forecasts', () => {
      expect(weatherCondition.mostSeriousCondition(null)).toBeUndefined();
      expect(weatherCondition.mostSeriousCondition(undefined)).toBeUndefined();
      expect(weatherCondition.mostSeriousCondition([])).toBeUndefined();
    });

    it('returns the most serious condition', () => {
      expect(
        weatherCondition.mostSeriousCondition([
          {
            date: new Date(),
            condition: 800,
            temperature: 287.15
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 286.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 289.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 290.65
          },
          {
            date: new Date(),
            condition: 801,
            temperature: 295.15
          },
          {
            date: new Date(),
            condition: 305,
            temperature: 297.15
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 295.65
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 293.65
          },
          {
            date: new Date(),
            condition: 800,
            temperature: 288.65
          }
        ])
      ).toEqual(305);
    });
  });
});


orig/src/components.d.ts
/* eslint-disable */
/* tslint:disable */
/**
 * This is an autogenerated file created by the Stencil compiler.
 * It contains typing information for all components that exist in this project.
 */
import { HTMLStencilElement, JSXBase } from '@stencil/core/internal';
import { ConditionIconPaths } from './models/condition-icon-paths';
import { Forecast } from './models/forecast';
export namespace Components {
  interface CsdemoCondition {
    bgColor: string;
    condition: number;
    iconPaths: ConditionIconPaths;
  }
  interface CsdemoDailyForecast {
    forecasts: Array<Forecast>;
    iconPaths: ConditionIconPaths;
    scale: string;
  }
  interface CsdemoTemperature {
    scale: string;
    temperature: number;
  }
  interface CsdemoUvIndex {
    uvIndex: number;
  }
}
declare global {
  interface HTMLCsdemoConditionElement extends Components.CsdemoCondition, HTMLStencilElement {}
  var HTMLCsdemoConditionElement: {
    prototype: HTMLCsdemoConditionElement;
    new (): HTMLCsdemoConditionElement;
  };
  interface HTMLCsdemoDailyForecastElement extends Components.CsdemoDailyForecast, HTMLStencilElement {}
  var HTMLCsdemoDailyForecastElement: {
    prototype: HTMLCsdemoDailyForecastElement;
    new (): HTMLCsdemoDailyForecastElement;
  };
  interface HTMLCsdemoTemperatureElement extends Components.CsdemoTemperature, HTMLStencilElement {}
  var HTMLCsdemoTemperatureElement: {
    prototype: HTMLCsdemoTemperatureElement;
    new (): HTMLCsdemoTemperatureElement;
  };
  interface HTMLCsdemoUvIndexElement extends Components.CsdemoUvIndex, HTMLStencilElement {}
  var HTMLCsdemoUvIndexElement: {
    prototype: HTMLCsdemoUvIndexElement;
    new (): HTMLCsdemoUvIndexElement;
  };
  interface HTMLElementTagNameMap {
    'csdemo-condition': HTMLCsdemoConditionElement;
    'csdemo-daily-forecast': HTMLCsdemoDailyForecastElement;
    'csdemo-temperature': HTMLCsdemoTemperatureElement;
    'csdemo-uv-index': HTMLCsdemoUvIndexElement;
  }
}
declare namespace LocalJSX {
  interface CsdemoCondition {
    bgColor?: string;
    condition?: number;
    iconPaths?: ConditionIconPaths;
  }
  interface CsdemoDailyForecast {
    forecasts?: Array<Forecast>;
    iconPaths?: ConditionIconPaths;
    scale?: string;
  }
  interface CsdemoTemperature {
    scale?: string;
    temperature?: number;
  }
  interface CsdemoUvIndex {
    uvIndex?: number;
  }
  interface IntrinsicElements {
    'csdemo-condition': CsdemoCondition;
    'csdemo-daily-forecast': CsdemoDailyForecast;
    'csdemo-temperature': CsdemoTemperature;
    'csdemo-uv-index': CsdemoUvIndex;
  }
}
export { LocalJSX as JSX };
declare module '@stencil/core' {
  export namespace JSX {
    interface IntrinsicElements {
      'csdemo-condition': LocalJSX.CsdemoCondition & JSXBase.HTMLAttributes<HTMLCsdemoConditionElement>;
      'csdemo-daily-forecast': LocalJSX.CsdemoDailyForecast & JSXBase.HTMLAttributes<HTMLCsdemoDailyForecastElement>;
      'csdemo-temperature': LocalJSX.CsdemoTemperature & JSXBase.HTMLAttributes<HTMLCsdemoTemperatureElement>;
      'csdemo-uv-index': LocalJSX.CsdemoUvIndex & JSXBase.HTMLAttributes<HTMLCsdemoUvIndexElement>;
    }
  }
}


src/components/csdemo-condition/csdemo-condition.e2e.ts
import { newE2EPage } from '@stencil/core/testing';

describe('csdemo-condition', () => {
  it('renders', async () => {
    const page = await newE2EPage();

    await page.setContent('<csdemo-condition></csdemo-condition>');
    const element = await page.find('csdemo-condition');
    const label = await page.find('csdemo-condition >>> .condition-label');
    expect(element).toHaveClass('hydrated');
    expect(label.textContent).toEqual('Unknown');
  });

  it('does not render an image if there is no image map', async () => {
    const page = await newE2EPage();

    await page.setContent('<csdemo-condition></csdemo-condition>');
    const element = await page.find('csdemo-condition');

    element.setProperty('condition', 200);
    await page.waitForChanges();
    const img = await page.find('csdemo-condition >>> img');
    expect(img).toBeFalsy();
  });

  it('renders the correct icon for the condition', async () => {
    const paths = {
      sunny: './assets/images/sunny.png',
      cloudy: './assets/images/cloudy.png',
      lightRain: './assets/images/light-rain.png',
      shower: './assets/images/shower.png',
      sunnyThunderStorm: './assets/images/sunny-tstorm.png',
      thunderStorm: './assets/images/tstorm.png',
      fog: './assets/images/fog.png',
      snow: './assets/images/snow.png',
      unknown: './assets/images/unknown.png'
    };

    const page = await newE2EPage();

    await page.setContent('<csdemo-condition></csdemo-condition>');
    const element = await page.find('csdemo-condition');

    element.setProperty('iconPaths', paths);
    element.setProperty('condition', 200);
    await page.waitForChanges();
    const img = await page.find('csdemo-condition >>> img');
    expect(img.getAttribute('src')).toEqual(paths.thunderStorm);

    element.setProperty('condition', 229);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.thunderStorm);

    element.setProperty('condition', 230);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.sunnyThunderStorm);

    element.setProperty('condition', 299);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.sunnyThunderStorm);

    element.setProperty('condition', 300);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.lightRain);

    element.setProperty('condition', 399);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.lightRain);

    element.setProperty('condition', 500);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.shower);

    element.setProperty('condition', 599);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.shower);

    element.setProperty('condition', 600);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.snow);

    element.setProperty('condition', 699);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.snow);

    element.setProperty('condition', 701);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.fog);

    element.setProperty('condition', 771);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.fog);

    element.setProperty('condition', 800);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.sunny);

    element.setProperty('condition', 801);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.cloudy);

    element.setProperty('condition', 809);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.cloudy);

    element.setProperty('condition', 810);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.unknown);

    element.setProperty('condition', 903);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.snow);

    element.setProperty('condition', 904);
    await page.waitForChanges();
    expect(img.getAttribute('src')).toEqual(paths.sunny);
  });

  it('renders the correct icon for the condition', async () => {
    const page = await newE2EPage();

    await page.setContent('<csdemo-condition></csdemo-condition>');
    const element = await page.find('csdemo-condition');

    element.setProperty('condition', 200);
    await page.waitForChanges();
    const label = await page.find('csdemo-condition >>> .condition-label');
    expect(label.textContent).toEqual('Thunderstorms');

    element.setProperty('condition', 229);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Thunderstorms');

    element.setProperty('condition', 230);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Scattered Storms');

    element.setProperty('condition', 299);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Scattered Storms');

    element.setProperty('condition', 300);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Rain');

    element.setProperty('condition', 399);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Rain');

    element.setProperty('condition', 500);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Showers');

    element.setProperty('condition', 599);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Showers');

    element.setProperty('condition', 600);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Snow');

    element.setProperty('condition', 699);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Snow');

    element.setProperty('condition', 701);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Fog');

    element.setProperty('condition', 771);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Fog');

    element.setProperty('condition', 800);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Sunny');

    element.setProperty('condition', 801);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Cloudy');

    element.setProperty('condition', 809);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Cloudy');

    element.setProperty('condition', 810);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Unknown');

    element.setProperty('condition', 903);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Snow');

    element.setProperty('condition', 904);
    await page.waitForChanges();
    expect(label.textContent).toEqual('Sunny');
  });
});
